File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int n = nums.size();
        int l = 0;
        int r = n - 1;

        vector<int> vect;

        while (l <= r) {
            if (abs(nums[l]) < abs(nums[r])) {
                vect.push_back(nums[r] * nums[r]);
                r--;
            } else {
                vect.push_back(nums[l] * nums[l]);
                l++;
            }
        }

        reverse(vect.begin(), vect.end());

        return vect;
    }
};

File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.py
class Solution:
    def sortedSquares(self, nums: list[int]) -> list[int]:
        # Get the length of the input array
        n = len(nums)
        
        # Initialize a result array of the same size to store squared values in sorted order
        result = [0] * n
        
        # Initialize two pointers: 'left' at the beginning and 'right' at the end of the array
        left = 0
        right = n - 1
        
        # Fill the result array from right to left (largest squares to smallest squares)
        for i in range(n - 1, -1, -1):
            # Compare the absolute values at the two pointers to find the larger square
            if abs(nums[left]) > abs(nums[right]):
                result[i] = nums[left] ** 2
                # Move the left pointer forward
                left += 1
            else:
                result[i] = nums[right] ** 2
                # Move the right pointer backward
                right -= 1
                
        # Return the newly sorted array of squares
        return result